mm2src/lp_ordermatch/lp_bot.rs

```diff
@@ -0,0 +1,168 @@
+//
+//  lp_bot.rs
+//  marketmaker
+//
+
+use common::{mm_ctx::{from_ctx, MmArc},
+             mm_number::MmNumber};
+use derive_more::Display;
+use futures::lock::Mutex as AsyncMutex;
+use std::time::{Duration, SystemTime, UNIX_EPOCH};
+use std::{collections::HashMap, sync::Arc};
+
+#[cfg(test)] use mocktopus::macros::*;
+
+#[path = "simple_market_maker.rs"] mod simple_market_maker_bot;
+pub use simple_market_maker_bot::{process_price_request, start_simple_market_maker_bot, stop_simple_market_maker_bot,
+                                  StartSimpleMakerBotRequest};
+
+#[cfg(all(test, not(target_arch = "wasm32")))]
+#[path = "simple_market_maker_tests.rs"]
+pub mod simple_market_maker_tests;
+
+#[derive(PartialEq)]
+enum TradingBotState {
+    Running,
+    Stopping,
+    Stopped,
+}
+
+impl Default for TradingBotState {
+    fn default() -> Self { TradingBotState::Stopped }
+}
+
+pub type SimpleMakerBotRegistry = HashMap<String, SimpleCoinMarketMakerCfg>;
+
+#[derive(Debug, Serialize, Deserialize, Display, Clone)]
+#[display(fmt = "{} {} {} {}", base, rel, enable, spread)]
+pub struct SimpleCoinMarketMakerCfg {
+    pub base: String,
+    pub rel: String,
+    #[serde(rename = "min_volume")]
+    pub min_volume_percentage: Option<MmNumber>,
+    pub spread: MmNumber,
+    pub base_confs: Option<u64>,
+    pub base_nota: Option<bool>,
+    pub rel_confs: Option<u64>,
+    pub rel_nota: Option<bool>,
+    pub enable: bool,
+    pub price_elapsed_validity: Option<f64>,
+    pub check_last_bidirectional_trade_thresh_hold: Option<bool>,
+    pub max: Option<bool>,
+    pub balance_percent: Option<MmNumber>,
+}
+
+#[derive(Default)]
+pub struct TickerInfosRegistry(HashMap<String, TickerInfos>);
+
+#[derive(Debug, Serialize, Deserialize)]
+pub struct TickerInfos {
+    ticker: String,
+    last_price: MmNumber,
+    last_updated: String,
+    last_updated_timestamp: u64,
+    #[serde(rename = "volume24h")]
+    volume24_h: MmNumber,
+    price_provider: Provider,
+    volume_provider: Provider,
+    #[serde(rename = "sparkline_7d")]
+    sparkline_7_d: Option<Vec<f64>>,
+    sparkline_provider: Provider,
+    #[serde(rename = "change_24h")]
+    change_24_h: MmNumber,
+    #[serde(rename = "change_24h_provider")]
+    change_24_h_provider: Provider,
+}
+
+#[derive(Debug, Serialize, Deserialize, Clone, PartialEq)]
+pub enum Provider {
+    #[serde(rename = "binance")]
+    Binance,
+    #[serde(rename = "coingecko")]
+    Coingecko,
+    #[serde(rename = "coinpaprika")]
+    Coinpaprika,
+    #[serde(rename = "unknown")]
+    Unknown,
+}
+
+impl Default for Provider {
+    fn default() -> Self { Provider::Unknown }
+}
+
+#[derive(Default)]
+pub struct TradingBotContext {
+    trading_bot_states: AsyncMutex<TradingBotState>,
+    trading_bot_cfg: AsyncMutex<SimpleMakerBotRegistry>,
+}
+
+#[derive(Default, Clone, Debug)]
+pub struct RateInfos {
+    base: String,
+    rel: String,
+    price: MmNumber,
+    last_updated_timestamp: u64,
```

Comment:
Here I use it as a `u64` because the API can return `0` for timestamp if it's invalid and I use it here:

https://github.com/KomodoPlatform/atomicDEX-API/blob/6c917c27906e986b5b4d82a1f839c70f852584a3/mm2src/lp_ordermatch/simple_market_maker.rs#L386

Reply:
If the value can have an uninitialized state it's more idiomatic to use `Option` instead of relying on the default magic constant.